### ISSUE TITLE
fastbuild has become slow

### ISSUE THREAD

**Issue by dae** [bazelbuild/rules_rust#1093]
I'm seeing a 5x slowdown building my crate with #1091.

Building with latest main:

   38.153 s   92.26%   action 'Compiling Rust rlib anki (198 files)'

After reverting #1091:

    7.646 s   83.10%   action 'Compiling Rust rlib anki (198 files)'

Building my crate in release mode is actually faster than opt level 1:

   26.012 s   92.79%   action 'Compiling Rust rlib anki (198 files)'

Perhaps my crate is an outlier here, but I thought I should mention it, as it's more than "very slightly" :-)

**Comment by UebelAndre** [comment]
Sadness! And I'm assuming caching is not a factor here?

**Comment by UebelAndre** [comment]
I've opened up a PR (https://github.com/bazelbuild/rules_rust/pull/1094) to revert the change while we discuss the details in this issue.

I think a build flag would be a good option. Something similar to this?
https://github.com/bazelbuild/rules_rust/blob/ea22478f6dac4604b4ef38fb8e713cd1c0880084/BUILD.bazel#L22-L26

**Comment by hlopko** [comment]
Yup build_settings is what I'd vote for as well. The numbers are suspricious, also the fact that opt-level=1 is slower than 2, maybe this is something we should open an issue for on rust-lang/rust

**Comment by mfarrugi** [comment]
These are just the default values, they can be overridden.
https://github.com/bazelbuild/rules_rust/blob/main/rust/toolchain.bzl#L321-L325
https://github.com/bazelbuild/rules_rust/blob/main/rust/toolchain.bzl#L343-L347

I think it makes sense to make the configurable toolchain parameters easier to pass through `rust_repositories` , regardless of whether we add more flag overrides.

**Comment by dae** [comment]
The tests I did above were on an arm64 Mac. I tried on an amd64 Linux machine, and things weren't as bad. It appears the Rust developers are aware of the issue, as L1 is no longer slower than L3 on the latest nightly on an arm64 machine:

```
CARGO_INCREMENTAL=0 CARGO_PROFILE_DEV_OPT_LEVEL=3 CARGO_PROFILE_DEV_DEBUG=false cargo +nightly build -Z timings
   Completed anki v0.0.0 in 19.0s

CARGO_INCREMENTAL=0 CARGO_PROFILE_DEV_OPT_LEVEL=1 CARGO_PROFILE_DEV_DEBUG=false cargo +nightly build -Z timings
   Completed anki v0.0.0 in 15.6s

CARGO_INCREMENTAL=0 CARGO_PROFILE_DEV_OPT_LEVEL=0 CARGO_PROFILE_DEV_DEBUG=false cargo +nightly build -Z timings
   Completed anki v0.0.0 in 7.4s
```

For reference, these are the numbers I see on nightly on an amd64 machine:

```
CARGO_INCREMENTAL=0 CARGO_PROFILE_DEV_OPT_LEVEL=3 CARGO_PROFILE_DEV_DEBUG=false cargo +nightly build -Z timings
   Completed anki v0.0.0 in 18.0s       

CARGO_INCREMENTAL=0 CARGO_PROFILE_DEV_OPT_LEVEL=1 CARGO_PROFILE_DEV_DEBUG=false cargo +nightly build -Z timings
   Completed anki v0.0.0 in 13.5s

CARGO_INCREMENTAL=0 CARGO_PROFILE_DEV_OPT_LEVEL=0 CARGO_PROFILE_DEV_DEBUG=false cargo +nightly build -Z timings
   Completed anki v0.0.0 in 8.3s
```

So even with these improvements in the pipeline, it looks like an opt level of 1 will mean a more than 2x increase in build time on arm64, and an approx 1.6x increase on amd64 for my use case. So an easy way to adjust the default levels with rust_repositories() or similar would be great. :-)

**Comment by liningpan** [comment]
Is it possible to set `opt-level` directly from `rust_register_toolchains`?
